fetchcode/package_registry/__init__.py

```diff
@@ -0,0 +1,154 @@
+import requests
+
+
+router = Router()
+
+
+@attrs
+class URLData:
+    """
+    Represent the Data extracted from any URL/PURL
+    `homepage_url` : URL to the homepage for this package
+    `api_url` :  Return the package repository API URL to obtain structured data for this package
+    `vcs_url` : URL for VCS download
+    `codeview_url` : URL where the code can be browsed online
+    `bugtracking_url` : URL to the issue or bug tracker for this package
+    `download_url` :  Return the package repository download URL to download the actual archive of code of this package. 
+    `documentation_url` : URL where the documentation can be found for this package
+    `readme_url` : URL where readme can be found for this package
+    `reverse_dependencies_url` : URL where reverse dependencies can be found for this package
+    `author_url` : URL where authors can be found regarding this package
+    `tags` : List of tags for this package
+    `tags_download_urls` : List of download URLS for tags of this package
+    `tags_readme_urls` : List of readme URLS for tags of this package
+    `tags_dependency_urls` : List of dependency URLs for tags of this package 
+    """
+    homepage_url = attrib(default=None)
+    api_url = attrib(default=None)
+    vcs_url = attrib(default=None)
+    codeview_url = attrib(default=None)
+    bugtracking_url = attrib(default=None)
+    download_url = attrib(default=None)
+    documentation_url = attrib(default=None)
+    readme_url = attrib(default=None)
+    reverse_dependencies_url = attrib(default=None)
+    author_url = attrib(default=None)
+    tags = attrib(default=None)
+    tags_download_urls = attrib(default=None)
+    tags_readme_urls = attrib(default=None)
+    tags_dependency_urls = attrib(default=None)
+
+
+def info(url):
+    """
+    Return data according to the `url` string
+    `url` string can be purl too
+    """
+    if url:
+        try:
+            return router.process(url)
+        except NoRouteAvailable:
+            return
+
+
+def get_response(url):
+    """
+    Return json object for a `url` string
+    """
+    resp = requests.get(url)
+    if "json" in dir(resp):
+        return resp.json()
+
+    raise Exception("Response of this URL does not contain json object")
+
+
+@router.route("pkg:cargo/.*")
+def cargo_data(purl):
+    """
+    Return `data` mapping of various URLs, from the `purl` string
+    """
+    purl_data = PackageURL.from_string(purl)
+    base_path = "https://crates.io"
+    name = purl_data.name
+    version = purl_data.version
+    api_url = f"{base_path}/api/v1/crates/{name}"
+    download_url = f"{api_url}/{version}/download"
+    readme_url = f"{api_url}/{version}/readme"
+    response = get_response(api_url)
+    crate = response.get("crate")
+    versions = response.get("versions", [])
+    homepage_url = None
+    documentation_url = None
+    codeview_url = None
+    reverse_dependencies_url = None
+    author_url = None
```

Comment:
Do we really need to declare these at the top if they are set to `None`? Shouldnt the attrs class take care of this?

Reply:
Every api will not give every type of url data, like every cargo may or may not contain codeview url, but in most of cases API returns codeview url, so it can be none or can be found, but like none of them gives bugtracking or VCS url, so we are not finding for it and it is handled by attributes in that case. If there is anything I am missing, I will like to have some suggestions.:)

Reply:
I guess my objection to this is lines 101-105 are just wasted variable initializations. Even if we set these values later, we do not need to initialize them to None

Reply:
Okay, I am getting you, I just thought that it might be a case when crate would not be in response, line 108 then these all won't be initiated at any point and will be directly called at line 142-154, so can you give me some suggestions regarding that :)

Reply:
I will adjust the code accordingly, and will write test cases for it and then begin working with npm from.now

Reply:
We can check inverse of line 107, then we can directly acess all of the URLs inside crate, what say, by this way we don't need to initialize 101-105, what say :)

Reply:
I tried the above thing, I was getting this when I ran tests when the crate was not in response `UnboundLocalError: local variable 'reverse_dependencies_url' referenced before assignment`, so I thought to initialize them to None only when crate will be None. suggestions on this :)